src/modules/flow/freegeoip/freegeoip.c

```diff
@@ -0,0 +1,239 @@
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "freegeoip-gen.h"
+#include "sol-flow-internal.h"
+
+#include <errno.h>
+#include <limits.h>
+#include <stdio.h>
+
+#include <sol-http-client.h>
+#include <sol-json.h>
+#include <sol-mainloop.h>
+#include <sol-util.h>
+
+struct freegeoip_data {
+    struct sol_flow_node *node;
+    char *endpoint;
+};
+
+#define RESPONSE_CHECK_API(response_, mdata_) \
+    do { \
+        if (unlikely(!response_)) { \
+            sol_flow_send_error_packet(mdata_->node, EINVAL, \
+                "Error while reaching Freegeoip"); \
+            return; \
+        } \
+        if (unlikely(response_->api_version != SOL_HTTP_RESPONSE_API_VERSION)) { \
+            SOL_ERR("Unexpected API version (response is %u, expected %u)", \
+                response->api_version, SOL_HTTP_RESPONSE_API_VERSION); \
+            return; \
+        } \
+    } while (0)
+
+enum fields {
+    FIELD_IP = 1<<0,
+    FIELD_COUNTRY_CODE = 1<<1,
+    FIELD_COUNTRY_NAME = 1<<2,
+    FIELD_REGION_CODE = 1<<3,
+    FIELD_CITY_NAME = 1<<4,
+    FIELD_ZIP_CODE = 1<<5,
+    FIELD_TIMEZONE = 1<<6,
+    FIELD_LATITUDE = 1<<7,
+    FIELD_LONGITUDE = 1<<8,
+    FIELD_ALL = 0x1ff
+};
+
+static bool
+json_token_to_double(const struct sol_json_token *token, double *output)
+{
+    char *str = strndupa(token->start, token->end - token->start);
+    char *endptr;
+    double value;
+
+    errno = 0;
+    value = strtod(str, &endptr);
+    if (errno != 0)
+        return false;
+    if (!endptr || *endptr)
+        return false;
+    *output = value;
+    return true;
+}
+
+static void
+freegeoip_query_finished(void *data, struct sol_http_response *response)
+{
+    struct freegeoip_data *mdata = data;
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+    enum fields fields = FIELD_ALL;
+    double latitude = 0.0f;
+    double longitude = 0.0f;
+
+    RESPONSE_CHECK_API(response, mdata);
+
+    if (!response->content.used) {
```

Comment:
shouldn't we send an error packet here? Maybe merge this in the following 'if' that checks for response code != 200?